File: src/main/kotlin/org/kite9/diagram/common/algorithms/fg/AbstractArc.java
package org.kite9.diagram.common.algorithms.fg;

import org.kite9.diagram.logging.LogicException;

abstract class AbstractArc implements Arc {

	protected int flow = 0;
	protected Node from;
	protected Node to;
	protected String label;

	public AbstractArc() {
		super();
	}

	public int getFlow() {
		return flow;
	}

	public int getFlowFrom(Node n) {
		if (n==from) {
			return flow;
		} else if (n==to) {
			return -flow;
		} else {
			throw new LogicException("Node not joined to arc");
		}
	}

	public void pushFlow(int flow) {
		this.flow += flow;
		from.pushFlow(-flow);
		to.pushFlow(flow);
	}

	public void setFlow(int flow) {
		this.flow = flow;
	}

	public Node getFrom() {
		return from;
	}

	public void setFrom(Node from) {
		this.from = from;
	}

	public Node getTo() {
		return to;
	}

	public void setTo(Node to) {
		this.to = to;
	}

	@Override
	public String toString() {
		return label;
	}

	public Node otherEnd(Node n) {
		if (from==n) {
			return to;
		}
		if (to==n)
			return from;
		return null;
	}

	@Override
	public int hashCode() {
		return label.hashCode();
	}

	private boolean blocked = false;
	
	@Override
	public void setBlocked(boolean blocked) {
		this.blocked = blocked;
	}

	@Override
	public boolean isBlocked() {
		return blocked;
	}
	
	@Override
	public String getID() {
		return label;
	}

}

File: src/main/kotlin/org/kite9/diagram/common/algorithms/fg/AbstractArc.kt
package org.kite9.diagram.common.algorithms.fg

import org.kite9.diagram.logging.LogicException

abstract class AbstractArc(override var from: Node, override var to: Node, val label: String, override var flow: Int = 0) : Arc {

    override fun getFlowFrom(n: Node): Int {
        return if (n === from) {
            flow
        } else if (n === to) {
            -flow
        } else {
            throw LogicException("Node not joined to arc")
        }
    }

    override fun pushFlow(flow: Int) {
        this.flow += flow
        from.pushFlow(-flow)
        to.pushFlow(flow)
    }

    override fun toString(): String {
        return label
    }

    override fun otherEnd(n: Node): Node {
        if (from === n) {
            return to
        }
        return if (to === n) from else throw IllegalArgumentException("Asking for end that's not set")
    }

    override fun hashCode(): Int {
        return label.hashCode()
    }

    override fun getID(): String {
        return label
    }

    init {
        from.arcs.add(this)
        to.arcs.add(this)
    }
}